helix-term/src/ui/editor.rs

```diff
@@ -1071,6 +1071,7 @@ impl EditorView {
 
     pub fn handle_idle_timeout(&mut self, cx: &mut commands::Context) -> EventResult {
         commands::compute_inlay_hints_for_all_views(cx.editor, cx.jobs);
+        commands::pull_diagnostic_for_current_doc(cx.editor, cx.jobs);
```

Comment:
Inlay hints were introduced before the event system that @pascalkuthe was mentioning in https://github.com/helix-editor/helix/pull/7900#issuecomment-1735911802. Pulling diagnostics should be done with an event system hook rather than here on idle timeout. I'm not sure we have a good PR to reference for how to do this yet. https://github.com/helix-editor/helix/pull/9668 and https://github.com/helix-editor/helix/pull/9638 show using hooks and #8021 introduced the event system initially and is probably the best background

Reply:
Thanks for the review.

Ah i see. I hoping this was related to the dispatch of DiagnosticsDidChange event. This makes more sense. 

Does it make sense to register a hook for DocumentDidChange which handles pulling and adding diagnostics?

Reply:
yeah that should do it you need an async handler to make sure they are debounced. There are a decent amount of hooks now that subscribe to `DocumentDidChange` so you can take a look at how those work. A good lsp related example is probably signature help (just that in your case you don't care about the selection changing).


I also don't think this should really live under commands at all. I don't think we need to support manually pulling diagnostics and if we do we can use a  handler for that (like we do for manually requesting completions). All related code should be (self-contained) inside the hook. This allows allows adding complexity like tracking result-id across multiple requests.

In a sense the most elegant way to refactor this would be to make is so that you decouple our current diagnostic handling from `publish-diagnostic` and then essentially create an internal pulbish diagnostic notification that can either come from the lsp directly or from our pull diagnostic handler.

Reply:
Thanks for clarification. 
I have pushed a refactoring with basically the same logic just with an async hook instead of idle timeout.

Let me know if i have missed something.

Now helix will only requesting diagnostics after document has changed. Is there a nice way to initiate the hook when the language server is started?